File: testing/web/tests/engine/core/types/LengthValue.test.ts
// Tests for extractLength — one test per documented IR shape (see examples/primitives/lengths-*.json).
import { describe, it, expect } from 'vitest';
import { extractLength, toCssLength } from '../../../../src/style/engine/core/types/LengthValue';

describe('extractLength', () => {
  // ---- absolute units (lengths-absolute.json) --------------------------------
  it('handles raw { px:N } shape with no wrapper', () => {
    expect(extractLength({ px: 20 })).toEqual({ kind: 'exact', px: 20 });
  });

  it('handles { type:"length", px:N } wrapper (Width case)', () => {
    expect(extractLength({ type: 'length', px: 200 })).toEqual({ kind: 'exact', px: 200 });
  });

  it('handles { type:"length", px:N, original:{v,u:"PT"} } — px wins over original', () => {
    expect(extractLength({ type: 'length', px: 160, original: { v: 120, u: 'PT' } }))
      .toEqual({ kind: 'exact', px: 160 });
  });

  // ---- font-relative (lengths-font-relative.json) ----------------------------
  it('handles font-relative em with no px (context-dependent)', () => {
    expect(extractLength({ type: 'length', original: { v: 2, u: 'EM' } }))
      .toEqual({ kind: 'relative', value: 2, unit: 'em' });
  });

  it('handles rem / rlh / cap / ch / ic / lh / ex units', () => {
    for (const u of ['REM','RLH','CAP','CH','IC','LH','EX']) {
      const out = extractLength({ type: 'length', original: { v: 3, u } });
      expect(out.kind).toBe('relative');
    }
  });

  // ---- viewport (lengths-viewport.json) --------------------------------------
  it('handles classic viewport units', () => {
    expect(extractLength({ type: 'length', original: { v: 50, u: 'VW' } }))
      .toEqual({ kind: 'relative', value: 50, unit: 'vw' });
  });

  it('handles small/large/dynamic viewport variants', () => {
    for (const u of ['SVW','LVH','DVMIN','SVMAX','LVI','DVB']) {
      const out = extractLength({ type: 'length', original: { v: 30, u } });
      expect(out.kind).toBe('relative');
    }
  });

  // ---- container-query (lengths-container.json) ------------------------------
  it('handles container-query units', () => {
    expect(extractLength({ type: 'length', original: { v: 50, u: 'CQW' } }))
      .toEqual({ kind: 'relative', value: 50, unit: 'cqw' });
    expect(extractLength({ type: 'length', original: { v: 30, u: 'CQMIN' } }).kind).toBe('relative');
  });

  // ---- intrinsic (lengths-intrinsic.json) ------------------------------------
  it('handles bare "auto" string', () => {
    expect(extractLength('auto')).toEqual({ kind: 'auto' });
  });

  it('handles bare "min-content" / "max-content" / "fit-content"', () => {
    expect(extractLength('min-content')).toEqual({ kind: 'intrinsic', intrinsicKind: 'min-content' });
    expect(extractLength('max-content')).toEqual({ kind: 'intrinsic', intrinsicKind: 'max-content' });
    expect(extractLength('fit-content')).toEqual({ kind: 'intrinsic', intrinsicKind: 'fit-content' });
  });

  // ---- percentage and fr (lengths-special.json) ------------------------------
  it('handles { type:"percentage", value:N }', () => {
    expect(extractLength({ type: 'percentage', value: 50 }))
      .toEqual({ kind: 'relative', value: 50, unit: 'percent' });
  });

  it('handles { fr:N } grid-fraction shape', () => {
    expect(extractLength({ fr: 2 })).toEqual({ kind: 'fraction', fr: 2 });
  });

  // ---- edge cases ------------------------------------------------------------
  it('returns unknown on null/undefined/empty', () => {
    expect(extractLength(null).kind).toBe('unknown');
    expect(extractLength(undefined).kind).toBe('unknown');
    expect(extractLength({}).kind).toBe('unknown');
  });

  it('returns unknown on unrecognised unit', () => {
    expect(extractLength({ original: { v: 5, u: 'NOPE' } }).kind).toBe('unknown');
  });

  it('treats bare number as percentage (Phase-2 margin/padding convention)', () => {
    // Margin_Percent_10 / Padding_Percent_10 fixtures emit bare JSON numbers for %.
    expect(extractLength(10)).toEqual({ kind: 'relative', value: 10, unit: 'percent' });
    expect(extractLength(0)).toEqual({ kind: 'relative', value: 0, unit: 'percent' });
  });

  it('handles calc expression shape ({type:"calc",expression:...})', () => {
    expect(extractLength({ type: 'calc', expression: '100% - 10px' }))
      .toEqual({ kind: 'calc', expression: '100% - 10px' });
  });

  it('handles {expr:"calc(...)"} shape (Phase-2 spacing survey)', () => {
    // Spacing fixtures emit {expr:"calc(10px + 5px)"}; outer calc() wrapper stripped.
    expect(extractLength({ expr: 'calc(10px + 5px)' }))
      .toEqual({ kind: 'calc', expression: '10px + 5px' });
  });

  it('handles {expr:"..."} without an outer calc() wrapper', () => {
    expect(extractLength({ expr: '100% - 2em' }))
      .toEqual({ kind: 'calc', expression: '100% - 2em' });
  });
});

describe('toCssLength', () => {
  it('formats every kind', () => {
    expect(toCssLength({ kind: 'exact', px: 20 })).toBe('20px');
    expect(toCssLength({ kind: 'relative', value: 50, unit: 'percent' })).toBe('50%');
    expect(toCssLength({ kind: 'relative', value: 2, unit: 'em' })).toBe('2em');
    expect(toCssLength({ kind: 'auto' })).toBe('auto');
    expect(toCssLength({ kind: 'intrinsic', intrinsicKind: 'min-content' })).toBe('min-content');
    expect(toCssLength({ kind: 'fraction', fr: 1 })).toBe('1fr');
    expect(toCssLength({ kind: 'calc', expression: '10px + 5%' })).toBe('calc(10px + 5%)');
    expect(toCssLength({ kind: 'unknown' })).toBe('auto');
  });
});

File: testing/Android/app/src/test/java/com/styleconverter/test/style/core/types/LengthValueTest.kt
package com.styleconverter.test.style.core.types

// Unit tests for extractLength — at least one per documented IR shape quirk.
// Each shape snippet is copy-pasted from examples/primitives/*.json output.

import kotlinx.serialization.json.Json
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class LengthValueTest {

    // Helper to keep test bodies small.
    private fun parse(s: String) = Json.parseToJsonElement(s)

    @Test fun `absolute px wrapped in type-length object`() {
        // Quirk 1: Width uses { type: "length", px: N }.
        val v = extractLength(parse("""{"type":"length","px":200.0}"""))
        assertEquals(LengthValue.Exact(200.0), v)
    }

    @Test fun `absolute raw px without type wrapper`() {
        // Quirk 1 (continued): PaddingTop emits { px: N }.
        val v = extractLength(parse("""{"px":8.0}"""))
        assertEquals(LengthValue.Exact(8.0), v)
    }

    @Test fun `non-px absolute unit normalized to px with original carried`() {
        // lengths-absolute.json: 120pt → px=160, original={v:120,u:PT}.
        val v = extractLength(parse("""{"type":"length","px":160.0,"original":{"v":120.0,"u":"PT"}}"""))
        assertEquals(LengthValue.Exact(160.0), v)
    }

    @Test fun `font-relative em has no px fallback`() {
        val v = extractLength(parse("""{"type":"length","original":{"v":2.0,"u":"EM"}}"""))
        assertEquals(LengthValue.Relative(2.0, LengthUnit.EM, null), v)
    }

    @Test fun `font-relative rem rlh lh cap ic ex ch all parse`() {
        for (u in listOf("REM", "LH", "RLH", "CAP", "IC", "EX", "CH")) {
            val v = extractLength(parse("""{"type":"length","original":{"v":1.5,"u":"$u"}}"""))
            assertEquals(LengthValue.Relative(1.5, LengthUnit.valueOf(u), null), v)
        }
    }

    @Test fun `classic viewport units parse`() {
        for (u in listOf("VW", "VH", "VMIN", "VMAX", "VI", "VB")) {
            val v = extractLength(parse("""{"type":"length","original":{"v":50.0,"u":"$u"}}"""))
            assertEquals(LengthValue.Relative(50.0, LengthUnit.valueOf(u), null), v)
        }
    }

    @Test fun `small large dynamic viewport units parse`() {
        // Covers the 18 new viewport units the spec calls out.
        val units = listOf(
            "SVW", "SVH", "SVMIN", "SVMAX", "SVI", "SVB",
            "LVW", "LVH", "LVMIN", "LVMAX", "LVI", "LVB",
            "DVW", "DVH", "DVMIN", "DVMAX", "DVI", "DVB",
        )
        for (u in units) {
            val v = extractLength(parse("""{"type":"length","original":{"v":10.0,"u":"$u"}}"""))
            assertEquals(LengthValue.Relative(10.0, LengthUnit.valueOf(u), null), v)
        }
    }

    @Test fun `container-query units parse`() {
        for (u in listOf("CQW", "CQH", "CQI", "CQB", "CQMIN", "CQMAX")) {
            val v = extractLength(parse("""{"type":"length","original":{"v":50.0,"u":"$u"}}"""))
            assertEquals(LengthValue.Relative(50.0, LengthUnit.valueOf(u), null), v)
        }
    }

    @Test fun `intrinsic auto is bare string data`() {
        // Quirk 3.
        val v = extractLength(parse(""""auto""""))
        assertEquals(LengthValue.Auto, v)
    }

    @Test fun `intrinsic min-content and max-content bare strings`() {
        assertEquals(
            LengthValue.Intrinsic(LengthValue.IntrinsicKind.MIN_CONTENT),
            extractLength(parse(""""min-content"""")),
        )
        assertEquals(
            LengthValue.Intrinsic(LengthValue.IntrinsicKind.MAX_CONTENT),
            extractLength(parse(""""max-content"""")),
        )
    }

    @Test fun `percentage sizing uses its own wrapper shape`() {
        // Quirk 2: { type: "percentage", value: 50.0 } NOT an IRLength.
        val v = extractLength(parse("""{"type":"percentage","value":50.0}"""))
        assertEquals(LengthValue.Relative(50.0, LengthUnit.PERCENT, null), v)
    }

    @Test fun `grid fraction is Fraction variant`() {
        // Quirk 4: fr is grid-track shape, not an IRLength.
        val v = extractLength(parse("""{"fr":2.0}"""))
        assertEquals(LengthValue.Fraction(2.0), v)
    }

    @Test fun `Q unit normalizes to px like other absolutes`() {
        // Absolute units always land in Exact even when the IR keeps original.
        val v = extractLength(parse("""{"type":"length","px":18.89763779527559,"original":{"v":80.0,"u":"Q"}}"""))
        assertTrue(v is LengthValue.Exact)
    }

    @Test fun `unknown unit string degrades to UNKNOWN`() {
        // Defensive: an unrecognized unit keeps the value but tags UNKNOWN.
        val v = extractLength(parse("""{"type":"length","original":{"v":3.0,"u":"PARSECS"}}"""))
        assertEquals(LengthValue.Relative(3.0, LengthUnit.UNKNOWN, null), v)
    }

    @Test fun `null input returns Unknown not null`() {
        assertEquals(LengthValue.Unknown, extractLength(null))
    }

    @Test fun `empty object returns Unknown`() {
        assertEquals(LengthValue.Unknown, extractLength(parse("""{}""")))
    }

    @Test fun `arbitrary non-keyword string returns Unknown`() {
        assertEquals(LengthValue.Unknown, extractLength(parse(""""nonsense"""")))
    }
}
